File: app/src/main/java/com/elementary/tasks/navigation/fragments/DayViewFragment.java
package com.elementary.tasks.navigation.fragments;

import android.app.AlarmManager;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.elementary.tasks.R;
import com.elementary.tasks.birthdays.CalendarPagerAdapter;
import com.elementary.tasks.birthdays.EventsPagerItem;
import com.elementary.tasks.core.utils.Configs;
import com.elementary.tasks.core.utils.Module;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.databinding.FragmentDayViewBinding;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

import androidx.viewpager.widget.ViewPager;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class DayViewFragment extends BaseCalendarFragment {

    private static final String DATE_KEY = "date";
    private static final String POS_KEY = "position";

    private FragmentDayViewBinding binding;
    private ArrayList<EventsPagerItem> pagerData = new ArrayList<>();

    public static DayViewFragment newInstance(long date, int position) {
        DayViewFragment pageFragment = new DayViewFragment();
        Bundle arguments = new Bundle();
        arguments.putLong(DATE_KEY, date);
        arguments.putInt(POS_KEY, position);
        pageFragment.setArguments(arguments);
        return pageFragment;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle intent = getArguments();
        if (intent != null) {
            dateMills = intent.getLong(DATE_KEY, 0);
        }
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.day_view_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_voice:
                if (getCallback() != null) {
                    getCallback().onVoiceAction();
                }
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentDayViewBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    private void updateMenuTitles() {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        if (dateMills != 0) {
            calendar.setTimeInMillis(dateMills);
        }
        String dayString = TimeUtil.getDate(calendar.getTimeInMillis());
        if (getCallback() != null) getCallback().onTitleChange(dayString);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getCallback() != null) {
            getCallback().onTitleChange(getString(R.string.calendar));
            getCallback().onFragmentSelect(this);
            getCallback().onMenuSelect(R.id.nav_day_view);
            getCallback().setClick(view -> showActionDialog(false));
        }
        loadData();
    }

    private void loadData() {
        initProvider();
        Calendar calendar = Calendar.getInstance();
        if (dateMills != 0) {
            calendar.setTimeInMillis(dateMills);
            showEvents(calendar.getTime());
        } else {
            calendar.setTimeInMillis(System.currentTimeMillis());
            showEvents(calendar.getTime());
        }
    }

    private void showEvents(Date date) {
        pagerData.clear();
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        int targetDay = calendar.get(Calendar.DAY_OF_MONTH);
        int targetMonth = calendar.get(Calendar.MONTH);
        int targetYear = calendar.get(Calendar.YEAR);

        calendar.setTimeInMillis(System.currentTimeMillis());

        int position = 0;
        int targetPosition = -1;
        calendar.add(Calendar.DAY_OF_MONTH, -100);
        while (position < Configs.MAX_DAYS_COUNT + 100) {
            int mDay = calendar.get(Calendar.DAY_OF_MONTH);
            int mMonth = calendar.get(Calendar.MONTH);
            int mYear = calendar.get(Calendar.YEAR);
            if (mDay == targetDay && mMonth == targetMonth && mYear == targetYear) {
                targetPosition = position;
                pagerData.add(new EventsPagerItem(position, 1, mDay, mMonth, mYear));
            } else {
                pagerData.add(new EventsPagerItem(position, 0, mDay, mMonth, mYear));
            }
            position++;
            calendar.setTimeInMillis(calendar.getTimeInMillis() + AlarmManager.INTERVAL_DAY);
        }
        CalendarPagerAdapter pagerAdapter = new CalendarPagerAdapter(Module.isJellyMR2() ? getChildFragmentManager() : getFragmentManager(), pagerData);
        try {
            binding.pager.setAdapter(pagerAdapter);
        } catch (IllegalStateException ignored) {
        }
        binding.pager.setOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int i, float v, int i2) {

            }

            @Override
            public void onPageSelected(int i) {
                EventsPagerItem item = pagerData.get(i);
                Calendar calendar1 = Calendar.getInstance();
                calendar1.set(item.getYear(), item.getMonth(), item.getDay());
                dateMills = calendar1.getTimeInMillis();
                updateMenuTitles();
            }

            @Override
            public void onPageScrollStateChanged(int i) {

            }
        });
        binding.pager.setCurrentItem(targetPosition);
        updateMenuTitles();
    }
}

File: app/src/main/java/com/elementary/tasks/navigation/fragments/DayViewFragment.kt
package com.elementary.tasks.navigation.fragments

import android.app.AlarmManager
import android.os.Bundle
import android.view.*
import androidx.viewpager.widget.ViewPager
import com.elementary.tasks.R
import com.elementary.tasks.birthdays.CalendarPagerAdapter
import com.elementary.tasks.birthdays.EventsPagerItem
import com.elementary.tasks.core.utils.Configs
import com.elementary.tasks.core.utils.Module
import com.elementary.tasks.core.utils.TimeUtil
import kotlinx.android.synthetic.main.fragment_day_view.*
import java.util.*

/**
 * Copyright 2016 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class DayViewFragment : BaseCalendarFragment() {

    private val pagerData = ArrayList<EventsPagerItem>()

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val intent = arguments
        if (intent != null) {
            dateMills = intent.getLong(DATE_KEY, 0)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.day_view_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.action_voice -> {
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_day_view, container, false)
    }

    private fun updateMenuTitles() {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = System.currentTimeMillis()
        if (dateMills != 0L) {
            calendar.timeInMillis = dateMills
        }
        val dayString = TimeUtil.getDate(calendar.timeInMillis)
        if (callback != null) callback?.onTitleChange(dayString)
    }

    override fun onResume() {
        super.onResume()
        if (callback != null) {
            callback?.onTitleChange(getString(R.string.calendar))
            callback?.onFragmentSelect(this)
            callback?.onMenuSelect(R.id.nav_day_view)
        }
        loadData()
    }

    private fun loadData() {
        initProvider()
        val calendar = Calendar.getInstance()
        if (dateMills != 0L) {
            calendar.timeInMillis = dateMills
            showEvents(calendar.time)
        } else {
            calendar.timeInMillis = System.currentTimeMillis()
            showEvents(calendar.time)
        }
    }

    private fun showEvents(date: Date) {
        pagerData.clear()
        val calendar = Calendar.getInstance()
        calendar.time = date
        val targetDay = calendar.get(Calendar.DAY_OF_MONTH)
        val targetMonth = calendar.get(Calendar.MONTH)
        val targetYear = calendar.get(Calendar.YEAR)

        calendar.timeInMillis = System.currentTimeMillis()

        var position = 0
        var targetPosition = -1
        calendar.add(Calendar.DAY_OF_MONTH, -100)
        while (position < Configs.MAX_DAYS_COUNT + 100) {
            val mDay = calendar.get(Calendar.DAY_OF_MONTH)
            val mMonth = calendar.get(Calendar.MONTH)
            val mYear = calendar.get(Calendar.YEAR)
            if (mDay == targetDay && mMonth == targetMonth && mYear == targetYear) {
                targetPosition = position
                pagerData.add(EventsPagerItem(position, 1, mDay, mMonth, mYear))
            } else {
                pagerData.add(EventsPagerItem(position, 0, mDay, mMonth, mYear))
            }
            position++
            calendar.timeInMillis = calendar.timeInMillis + AlarmManager.INTERVAL_DAY
        }
        val pagerAdapter = CalendarPagerAdapter(if (Module.isJellyMR2) childFragmentManager else fragmentManager!!, pagerData)
        try {
            pager.adapter = pagerAdapter
        } catch (ignored: IllegalStateException) {
        }

        pager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(i: Int, v: Float, i2: Int) {

            }

            override fun onPageSelected(i: Int) {
                val item = pagerData[i]
                val calendar1 = Calendar.getInstance()
                calendar1.set(item.year, item.month, item.day)
                dateMills = calendar1.timeInMillis
                updateMenuTitles()
            }

            override fun onPageScrollStateChanged(i: Int) {

            }
        })
        pager.currentItem = targetPosition
        updateMenuTitles()
    }

    companion object {

        private const val DATE_KEY = "date"
        private const val POS_KEY = "position"

        fun newInstance(date: Long, position: Int): DayViewFragment {
            val pageFragment = DayViewFragment()
            val arguments = Bundle()
            arguments.putLong(DATE_KEY, date)
            arguments.putInt(POS_KEY, position)
            pageFragment.arguments = arguments
            return pageFragment
        }
    }
}
